astropy/io/misc/asdf/tags/coordinates/skycoord.py

```diff
@@ -0,0 +1,37 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+
+import numpy as np
+
+from asdf.yamlutil import custom_tree_to_tagged_tree, tagged_tree_to_custom_tree
+
+from astropy.coordinates import SkyCoord
+
+from ...types import AstropyType
+
+
+class SkyCoordType(AstropyType):
+    name = 'coordinates/skycoord'
+    types = [SkyCoord]
+    version = "1.0.0"
+
+    @classmethod
+    def to_tree(cls, obj, ctx):
+        return custom_tree_to_tagged_tree(obj.info._represent_as_dict(), ctx)
+
+    @classmethod
+    def from_tree(cls, tree, ctx):
+        return SkyCoord.info._construct_from_dict(tree)
+
+    @classmethod
+    def assert_equal(cls, old, new):
+        old_dict = old.info._represent_as_dict()
```

Comment:
Another way to do this check is with this function to directly compare the `SkyCoord` objects:

https://github.com/astropy/astropy/blob/de3933688433e22765b1bdffa175fcd954d691c3/astropy/table/tests/test_operations.py#L24

Maybe do both?  Comparing only the `_represent_as_dict()` bits assumes that this representation is complete.  I think that's true and tested elsewhere, but still...

Reply:
So when using `skycoord_equal` I actually get a few test failures (which seems to indicate that it's a better test!):

```python
c = SkyCoord(ra=1*u.deg, dec=2*u.deg, radial_velocity=10*u.km/u.s)
...
old = <SkyCoord (ICRS): (ra, dec) in deg
    (1., 2.)
 (radial_velocity) in km / s
    (10.,)>
new = <SkyCoord (ICRS): (ra, dec) in deg
    (1., 2.)>
```

It looks like in this case `radial_velocity` does not get serialized (almost typed "cerealized"... I guess it's breakfast time).

```python
c = SkyCoord(ra=1*u.deg, dec=2*u.deg, pm_ra_cosdec=2*u.mas/u.yr, pm_dec=1*u.mas/u.yr)
...
old = <SkyCoord (ICRS): (ra, dec) in deg
    (1., 2.)
 (pm_ra_cosdec, pm_dec) in mas / yr
    (2., 1.)>
new = <SkyCoord (ICRS): (ra, dec) in deg
    (1., 2.)>
```

In this case, `pm_ra_cosdec` and `pm_dec` don't appear to be serialized.

Since the `asdf` tag is just using the representation from `.info`, I think this problem might be elsewhere. @taldcroft I can dig deeper, though, unless it's obvious to you how to fix it.

Reply:
Yes, I was involved and am not sure why it doesn't work. But clearly it should since this means that coordinates with velocities almost certainly do not roundtrip to fits either.

Reply:
meant "should *be made to work* since ..." 

It may have something to do with the fact that the attributes are not guaranteed to be present.

Reply:
As @drdavella already noticed (I think), it looks like `_represent_as_dict` needs to be updated to look not only at `representation_component_names` but also `get_representation_component_names('s')` (checking there is in fact a derivative, and verifying, like for `Distance`, whether they are in fact defined). Ideally, we do this more elegantly...